# LAQN Updated Active sites/species File Function:
- laqn_remove notebook got slower so i will move the update function here.
- Start with paths and modules to import.

In [1]:
import pandas as pd
from pathlib import Path
import re
import glob

# paths beloww
base_dir = Path("/Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels")
# the metadata file for nan @value 
nanValue_path = base_dir / "data" / "laqn" / "missing" / "logs_nan_value.csv"

# Month abbreviation list for reference/use in functions
month_list = ['jan', 'feb', 'mar', 'apr', 'may', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec']

# analyse_affected_sites_2023 path below.
check_nonActive_path = base_dir / "data" / "laqn" / "missing" / "notActive_siteSpecies.csv"
output_notActive_siteSpecies_2023 = base_dir / "data" / "laqn" / "missing" / "notActive_siteSpecies_2023.csv"

# checks the removed site-species combinations against active list and removes them
existing_nonactive_path = base_dir / "data" / "laqn" / "missing" / "notActive_site_species.csv"

# analyse_affected_sites_2024 path below.
output_notActive_siteSpecies_2024 = base_dir / "data" / "laqn" / "missing" / "notActive_siteSpecies_2024.csv"

# analyse_affected_sites_2025 path below.
output_notActive_siteSpecies_2025 = base_dir / "data" / "laqn" / "missing" / "notActive_siteSpecies_2025.csv"

#calculating new issue rate, so I will be count all files in optimased folder and than recalculate the issue rate taking 
# out the notAcvtive_2024/2023 files from output_notActive_siteSpecies_2023, output_notActive_siteSpecies_2024 csv's
optimased_root = base_dir / "data" / "laqn" / "optimased"
log_file = base_dir / "data" / "laqn" / "missing" / "logs_missin_value.csv"

# Collect all CSV files in the optimased_root directory
all_csv_files = list(Path(optimased_root).rglob("*.csv"))
all_csv_filenames = set(f.name for f in all_csv_files)

# logs of removed files
log_csv_path = base_dir / "data" / "laqn" / "missing" / "logs_rm_notActive_23_24.csv"

# metadata path for active site-species
active_metadata_path=base_dir / "data" / "laqn" / "updated_actv_siteSpecies.csv"

## 1) analyse_affected_sites_2023 function to identify site/species with all months missing in 2023
    what does function:

- scans the missing value log for site/species combinations that have 100% missing values for all 12 months in 2023
- if an existing non-active site/species list is provided, only new combinations not already in that list are included
- prints the results and saves them as a CSV file for further review

    paths:

- nanValue_path: base_dir / data / laqn / missing / logs_nan_value.csv
- output_notActive_siteSpecies_2023: base_dir / data / laqn / missing / notActive_siteSpecies_2023.csv
- existing_nonactive_path (optional): base_dir / data / laqn / missing / notActive_site_species.csv
- new csv file save as: notActive_siteSpecies_2023.csv

In [40]:
def analyse_affected_sites_year(
    nanValue_path,  # value_100filtered_missing.csv
    output_notActive_site_species,  # recommended output path
    check_nonActive_path=None,  # notActive_site_species.csv
    year=2025,
    optimased_dir=None  # Optional: base dir for constructing file paths
):
    """
    1. Checks value_100filtered_missing.csv for site/species combos with 100% missing values for all months in the given year.
    2. Compares to notActive_site_species.csv and finds new combos not already listed.
    3. Adds actual filenames (and optionally full paths) for each combo from logs_nan_value.csv.
    4. Prints and saves these new combos to output_notActive_site_species.
    """

    def extract_month_number(month_str):
        month_map = {abbr: i+1 for i, abbr in enumerate(month_list)}
        m = re.match(r"\d{4}_(\w{3})", str(month_str).lower())
        if m:
            return month_map.get(m.group(1), None)
        return None

    # Load and filter for year, 100% missing
    df = pd.read_csv(nanValue_path, encoding='utf-8')
    df['month_number'] = df['month'].apply(extract_month_number)
    df = df[df['month_number'].notna()].copy()
    df_year = df[df['year'] == year]
    summary_year = df_year.groupby(['siteCode', 'SpeciesCode'])['month_number'].nunique().reset_index()
    affected_year = summary_year[summary_year['month_number'] == 12].copy()

    print(f"\nTotal site/species combos with 100% missing in {year}: {len(affected_year)}")
    print(affected_year)

    # Merge to get actual filenames and (optionally) full paths from nanValue_path
    merged = pd.merge(
        affected_year,
        df_year[['siteCode', 'SpeciesCode', 'filename', 'path']].drop_duplicates(),
        on=['siteCode', 'SpeciesCode'],
        how='left'
    )

    # Add expected filename prefix for reference
    def make_filename(row):
        return f"{str(row['siteCode']).lower()}_{str(row['SpeciesCode']).lower()}_"
    merged['expected_filename_prefix'] = merged.apply(make_filename, axis=1)
    if optimased_dir is not None:
        merged['expected_path_prefix'] = merged['expected_filename_prefix'].apply(
            lambda x: str(Path(optimased_dir) / x)
        )

    #  Compare to notActive_site_species.csv 
    if check_nonActive_path is not None:
        try:
            existing = pd.read_csv(check_nonActive_path, encoding='utf-8')
            existing_set = set(zip(existing['siteCode'], existing['SpeciesCode']))
            affected_year_set = set(zip(merged['siteCode'], merged['SpeciesCode']))
            new_combos = affected_year_set - existing_set
            new_affected_year = merged[
                merged.apply(lambda row: (row['siteCode'], row['SpeciesCode']) in new_combos, axis=1)
            ]
            print(f"\nNew site/species combos NOT in notActive_site_species.csv: {len(new_affected_year)}")
            print(new_affected_year)
            # Save to output
            new_affected_year.to_csv(output_notActive_site_species, index=False, encoding='utf-8')
            print(f"\nSaved new combos to: {output_notActive_site_species}")
            return new_affected_year
        except Exception as e:
            print(f"Warning: Could not filter by notActive_site_species.csv: {e}")
    else:
        print(f"\nNo notActive_site_species.csv provided for comparison.")
    merged.to_csv(output_notActive_site_species, index=False, encoding='utf-8')
    return merged


In [41]:
# Run the function for 2023
affected_2023 = analyse_affected_sites_year(
    nanValue_path,
    output_notActive_siteSpecies_2023,
    check_nonActive_path=existing_nonactive_path,
    year=2023
)

# Run the function for 2024
affected_2024 = analyse_affected_sites_year(
    nanValue_path,
    output_notActive_siteSpecies_2024,
    check_nonActive_path=existing_nonactive_path,
    year=2024
)

# Run the function for 2025
affected_2025 = analyse_affected_sites_year(
    nanValue_path,
    output_notActive_siteSpecies_2025,
    check_nonActive_path=existing_nonactive_path,
    year=2025
)



Total site/species combos with 100% missing in 2023: 41
    siteCode SpeciesCode  month_number
40       EN5       PM2.5            12
61       GT1         NO2            12
62       GT1        PM10            12
63       GT1       PM2.5            12
65       HG1       PM2.5            12
80       IS2       PM2.5            12
83       IS6       PM2.5            12
86       KF1        PM25            12
98       ME2         NO2            12
99       ME2        PM10            12
100      ME2       PM2.5            12
101      MEA         NO2            12
102      MEA       PM2.5            12
103      MEB        PM10            12
104      MEB       PM2.5            12
105      MR8        PM25            12
110      RHI         NO2            12
111      RHI        PM10            12
112      RHI       PM2.5            12
117      RI2       PM2.5            12
121      TH2       PM2.5            12
124      TH5         NO2            12
125      TH5        PM10            12
126    

## 2) calculate_adjusted_issue_rate
    This function calculates the adjusted issue rate of files with more than 20% missing @Value in the optimased directory, after excluding files associated with site/species combinations listed in the notActive_siteSpecies_2023.csv and notActive_siteSpecies_2024.csv files.

- How it works:

    - Reads the log file (logs_missin_value.csv) to count the number of files with >20% missing @Value.
    - Recursively counts all CSV files in the optimased directory.
    - Excludes from the total any files that match site/species pairs found in the not-active lists for 2023 and 2024.
    - Calculates the issue rate as: issue rate = ( Number of files with >20% missing @Value / total number of files) × 100


- Purpose:

    - To provide a more accurate data quality metric by removing files that are already flagged as not active for 2023 and 2024, so the issue rate reflects only the remaining, relevant files.

In [52]:
def detailed_issue_rate_excluding_notactive(optimased_dir, notactive_files):
    """
    Checks all files in optimased_dir for >20% missing @Value, and calculates the issue rate
    after excluding files listed in notActive_siteSpecies_2023/2024 (by filename).
    Returns: (issue_rate, total_files_checked, files_with_high_missing, missing_values_log)
    """

    # Collect filenames to exclude from notactive_files
    exclude_filenames = set()
    for naf in notactive_files:
        if Path(naf).exists():
            df_exclude = pd.read_csv(naf)
            if 'filename' in df_exclude.columns:
                exclude_filenames.update(df_exclude['filename'].dropna().astype(str).str.strip())
            else:
                for _, row in df_exclude.iterrows():
                    site = str(row['siteCode'])
                    species = str(row['SpeciesCode'])
                    pattern = f"{site}_{species}_"
                    matches = [fname for fname in all_csv_filenames if fname.startswith(pattern)]
                    exclude_filenames.update(matches)

    # Filter files to check (exclude notactive)
    files_to_check = [f for f in all_csv_files if f.name not in exclude_filenames]
    total_files_checked = len(files_to_check)
    files_with_high_missing = 0
    missing_values_log = []

    for filepath in files_to_check:
        try:
            df = pd.read_csv(filepath)
            if '@Value' in df.columns and len(df) > 0:
                missing_values = df['@Value'].isna().sum()
                empty_value_percentage = (100 * missing_values / len(df))
                if empty_value_percentage > 20:
                    files_with_high_missing += 1
                    missing_values_log.append({
                        'filename': filepath.name,
                        'path': str(filepath),
                        'EmptyValuePercentage': round(empty_value_percentage, 2)
                    })
        except Exception as e:
            print(f"Error reading {filepath}: {e}")

    issue_rate = (files_with_high_missing / total_files_checked * 100) if total_files_checked > 0 else 0.0
    print(f"Total files checked: {total_files_checked}")
    print(f"Files with >20% missing @Value: {files_with_high_missing}")
    print(f"Issue rate (excluding not-active): {issue_rate:.2f}%")
    print(f"Files excluded (not-active): {len(exclude_filenames)}")
    return issue_rate, total_files_checked, files_with_high_missing, missing_values_log




In [53]:
# Example usage:
notactive_files = [output_notActive_siteSpecies_2023, output_notActive_siteSpecies_2024]
issue_rate, total_files_checked, files_with_high_missing, missing_values_log = detailed_issue_rate_excluding_notactive(
    optimased_root, notactive_files
)
print(f"\nFinal issue rate (excluding not-active): {issue_rate:.2f}%")

Total files checked: 6201
Files with >20% missing @Value: 1628
Issue rate (excluding not-active): 26.25%
Files excluded (not-active): 1248

Final issue rate (excluding not-active): 26.25%


## 3) remove the notActive_siteSpecies_2023 and 2024 files from optimased dataset.
    - Total files checked: 6201
    - Files with >20% missing @Value: 1628
    - Issue rate (excluding not-active): 26.25%
    - Files excluded (not-active): 1248

    - Final issue rate (excluding not-active): 26.25%

In [ ]:
# Remove files using the exclusion logic from detailed_issue_rate_excluding_notactive and log removals

def rm_files_notactive(optimased_root, notactive_files, dry_run=True, log_csv_path=None):
    """
    Remove (or list for removal) all files in optimased_dir that would be excluded by detailed_issue_rate_excluding_notactive.
    If dry_run is True, only print the files that would be removed. If False, actually delete them.
    If log_csv_path is provided, save a CSV log of removed files with columns:
    siteCode,SpeciesCode,month_number,filename,path,expected_filename_prefix
    Returns the list of files removed (or would be removed).
    """
    exclude_filenames = set()
    file_info = {}
    # Build exclusion set and info from notactive_files
    for naf in notactive_files:
        if Path(naf).exists():
            df_exclude = pd.read_csv(naf)
            if 'filename' in df_exclude.columns:
                for _, row in df_exclude.iterrows():
                    fname = str(row['filename']).strip()
                    exclude_filenames.add(fname)
                    file_info[fname] = {
                        'siteCode': row.get('siteCode', ''),
                        'SpeciesCode': row.get('SpeciesCode', ''),
                        'month_number': row.get('month_number', ''),
                        'filename': fname,
                        'path': row.get('path', ''),
                        'expected_filename_prefix': row.get('expected_filename_prefix', '')
                    }
            else:
                for _, row in df_exclude.iterrows():
                    site = str(row['siteCode'])
                    species = str(row['SpeciesCode'])
                    pattern = f"{site}_{species}_"
                    for f in all_csv_files:
                        if f.name.startswith(pattern):
                            exclude_filenames.add(f.name)
                            file_info[f.name] = {
                                'siteCode': site,
                                'SpeciesCode': species,
                                'month_number': row.get('month_number', ''),
                                'filename': f.name,
                                'path': str(f),
                                'expected_filename_prefix': pattern
                            }
    files_to_remove = [f for f in all_csv_files if f.name in exclude_filenames]
    print(f"Found {len(files_to_remove)} files to remove from {optimased_root}.")
    log_rows = []
    if dry_run:
        print("Dry run: No files will be deleted. Files that would be removed:")
        for f in files_to_remove:
            print(f)
            if f.name in file_info:
                log_rows.append(file_info[f.name])
            else:
                log_rows.append({
                    'siteCode': '', 'SpeciesCode': '', 'month_number': '',
                    'filename': f.name, 'path': str(f), 'expected_filename_prefix': ''
                })
    else:
        for f in files_to_remove:
            try:
                f.unlink()
                print(f"Deleted: {f}")
            except Exception as e:
                print(f"Error deleting {f}: {e}")
            if f.name in file_info:
                log_rows.append(file_info[f.name])
            else:
                log_rows.append({
                    'siteCode': '', 'SpeciesCode': '', 'month_number': '',
                    'filename': f.name, 'path': str(f), 'expected_filename_prefix': ''
                })
    # adding to missing log
    if log_csv_path and log_rows:
        df_log = pd.DataFrame(log_rows)
        # df_log = df_log[['siteCode','SpeciesCode','month_number','filename','path','expected_filename_prefix']]
        # df_log.to_csv(log_csv_path, index=False, encoding='utf-8')
        print(f"Log written to {log_csv_path}")
    return files_to_remove




In [ ]:
# Example usage (dry run):
notactive_files = [output_notActive_siteSpecies_2023, output_notActive_siteSpecies_2024]
#for removal dry_run= False 
files_to_remove = rm_files_notactive(optimased_root, notactive_files, dry_run=True, log_csv_path=log_csv_path)
print(f"Total files that would be removed: {len(files_to_remove)}")

Found 1248 files to remove from /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased.
Error deleting /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2023_mar/MEA_PM2.5_2023-03-01_2023-03-31.csv: [Errno 2] No such file or directory: '/Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2023_mar/MEA_PM2.5_2023-03-01_2023-03-31.csv'
Error deleting /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2023_mar/IS6_PM2.5_2023-03-01_2023-03-31.csv: [Errno 2] No such file or directory: '/Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2023_mar/IS6_PM2.5_2023-03-01_2023-03-31.csv'
Error deleting /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2023_mar/WA9_NO2_2023-03-01_2023-03-31.csv: [Errno 2] No such file or directory: '/Users/bur

        Log written to /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/missing/logs_rm_notActive_23_24.csv
        Total files that would be removed: 1248

### 4)  remove all KF1 and MR8 PM2.5/PM25 files from the optimased directory:
 -  noticed that those files still at optimased path.
 - Below the function for remove them all.


In [10]:
def remove_kf1_mr8_pm25_files(optimased_root, dry_run=False):
    """
    Remove all KF1 and MR8 PM2.5/PM25 files from the optimased directory (all months/years).
    If dry_run is True, only print the files that would be removed. If False, actually delete them.
    """

    patterns = [
        "KF1_PM2.5_*.csv", "KF1_PM25_*.csv",
        "MR8_PM2.5_*.csv", "MR8_PM25_*.csv"
    ]
    files_to_rm = []
    for pattern in patterns:
        files_to_rm.extend(Path(optimased_root).rglob(pattern))
    print(f"Found {len(files_to_rm)} KF1/MR8 PM2.5/PM25 files to remove in {optimased_root}.")
    if dry_run:
        print("Dry run: No files will be deleted. Files that would be removed:")
        for f in files_to_rm:
            print(f)
    else:
        for f in files_to_rm:
            try:
                f.unlink()
                print(f"Deleted: {f}")
            except Exception as e:
                print(f"Error deleting {f}: {e}")
    print(f"Total files processed: {len(files_to_rm)}")
    return files_to_rm
    

In [11]:
removed_files = remove_kf1_mr8_pm25_files(optimased_root, dry_run=False)

# Log the removed files to the same log as not-active removals
if removed_files:
    import pandas as pd
    log_rows = []
    for f in removed_files:
        log_rows.append({
            'siteCode': '',
            'SpeciesCode': '',
            'month_number': '',
            'filename': f.name,
            'path': str(f),
            'expected_filename_prefix': f.name[:f.name.find('_', 5)] if '_' in f.name[5:] else ''
        })
    try:
        # If the log file exists, append; otherwise, create
        try:
            df_log = pd.read_csv(log_csv_path)
            df_new = pd.DataFrame(log_rows)
            df_log = pd.concat([df_log, df_new], ignore_index=True)
        except FileNotFoundError:
            df_log = pd.DataFrame(log_rows)
        df_log.to_csv(log_csv_path, index=False, encoding='utf-8')
        print(f"Appended {len(log_rows)} entries to log: {log_csv_path}")
    except Exception as e:
        print(f"Error writing to log: {e}")
else:
    print("No files were removed.")



Found 22 KF1/MR8 PM2.5/PM25 files to remove in /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased.
Deleted: /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2025_feb/KF1_PM25_2025-02-01_2025-02-28.csv
Deleted: /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2025_aug/KF1_PM25_2025-08-01_2025-08-31.csv
Deleted: /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2025_mar/KF1_PM25_2025-03-01_2025-03-31.csv
Deleted: /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2025_jul/KF1_PM25_2025-07-01_2025-07-31.csv
Deleted: /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2025_oct/KF1_PM25_2025-10-01_2025-10-31.csv
Deleted: /Users/burdzhuchaglayan/Desktop/data science projects/air-pollution-levels/data/laqn/optimased/2025_sep/KF1_

## 4) Analysis of Site-Species Overlaps(2023/2024),  Issue Rate Calculation

This function checks  which site/species combo appear in both the non-active lists. (2023 and 2024) and the optimised dataset. The goal is to understand which files can be 
removed and what the new issue rate will be after exclusion.
Because I realised that, the non active site/species not balanced in each year and trying to do machine learning won't be work, in this unbalanced data. Soo I need to balanced the data. 

    - Steps:
        - 1. Load the non-active site/species lists for 2023 and 2024
        - 2. Identify which combinations exist in the optimised folder
        - 3. Calculate overlap between non-active lists and actual files
        - 4. Determine new metadata after removal
        - 5. Calculate updated issue rate excluding non-active combinations

In [2]:
def compare_nonactive_with_files(
    notactive_2023_path,
    notactive_2024_path,
    optimised_root,
    active_metadata_path
):
    """
    Compares non-active site/species lists with files in the optimised folder without rm anything.
    
    Args:
        notactive_2023_path: Path to notActive_siteSpecies_2023.csv
        notactive_2024_path: Path to notActive_siteSpecies_2024.csv
        optimised_root: Root directory containing optimised CSV files
        active_metadata_path: Path to current active site/species metadata
        
    Returns:
        dict containing comparison results
    """
    
    # Load nonActive lists
    df_2023 = pd.read_csv(notactive_2023_path)
    df_2024 = pd.read_csv(notactive_2024_path)
    
    # Load current active metadata
    df_active = pd.read_csv(active_metadata_path)
    
    # Get all CSV files in optimised folder
    all_files = list(Path(optimised_root).rglob("*.csv"))
    
    # Extract site/species from filenames
    def extract_site_species(filename):
        parts = filename.stem.split('_')
        if len(parts) >= 2:
            return (parts[0].upper(), parts[1].upper())
        return None
    
    file_combinations = {}
    for f in all_files:
        combo = extract_site_species(f)
        if combo:
            if combo not in file_combinations:
                file_combinations[combo] = []
            file_combinations[combo].append(f.name)
    
    # Create sets of non-Active combinations
    nonactive_2023_set = set(zip(df_2023['siteCode'], df_2023['SpeciesCode']))
    nonactive_2024_set = set(zip(df_2024['siteCode'], df_2024['SpeciesCode']))
    all_nonactive = nonactive_2023_set | nonactive_2024_set
    
    # Find overlaps
    overlapping = {combo: files for combo, files in file_combinations.items() if combo in all_nonactive}
    non_overlapping = all_nonactive - set(file_combinations.keys())
    
    # Count total files that would be affected
    total_files_affected = sum(len(files) for files in overlapping.values())
    
    # Calculate what would remain
    active_combinations = set(zip(df_active['SiteCode'], df_active['SpeciesCode']))
    new_active_combinations = active_combinations - all_nonactive
    
    new_active_metadata = df_active[
        df_active.apply(
            lambda row: (row['SiteCode'], row['SpeciesCode']) in new_active_combinations,
            axis=1
        )
    ]
    
    # Print detailed comparison
    print("Comparison Analysis Report Non-Active Lists/ Optimised Files")
    print("=" * 40)
    
    print(f"\nCurrent State:")
    print(f"  Total files in optimised folder: {len(all_files)}")
    print(f"  Active site/species combinations: {len(active_combinations)}")
    
    print(f"\nNon-Active Lists:")
    print(f"  2023 combinations: {len(nonactive_2023_set)}")
    print(f"  2024 combinations: {len(nonactive_2024_set)}")
    print(f"  Combined unique combinations: {len(all_nonactive)}")
    
    print(f"\nOverlap Analysis:")
    print(f"  Non-active combinations with files: {len(overlapping)}")
    print(f"  Total files affected: {total_files_affected}")
    print(f"  Non-active combinations without files: {len(non_overlapping)}")
    
    print(f"\nProjected State After Removal:")
    print(f"  Files remaining: {len(all_files) - total_files_affected}")
    print(f"  Active combinations remaining: {len(new_active_combinations)}")
    print(f"  Combinations removed: {len(active_combinations) - len(new_active_combinations)}")
    
    print(f"\nData Loss Percentage:")
    file_loss = (total_files_affected / len(all_files) * 100) if len(all_files) > 0 else 0
    combo_loss = ((len(active_combinations) - len(new_active_combinations)) / len(active_combinations) * 100) if len(active_combinations) > 0 else 0
    print(f"  Files: {file_loss:.2f}%")
    print(f"  Site/species combinations: {combo_loss:.2f}%")
    
    return {
        'overlapping_combinations': overlapping,
        'non_overlapping_combinations': non_overlapping,
        'total_files_affected': total_files_affected,
        'new_active_metadata': new_active_metadata,
        'original_file_count': len(all_files),
        'remaining_file_count': len(all_files) - total_files_affected,
        'original_combo_count': len(active_combinations),
        'remaining_combo_count': len(new_active_combinations)
    }

In [4]:
# run comparison

result = compare_nonactive_with_files(notactive_2023_path=output_notActive_siteSpecies_2023,
    notactive_2024_path=output_notActive_siteSpecies_2024,
    optimised_root=optimased_root,
    active_metadata_path= active_metadata_path
)

Comparison Analysis Report Non-Active Lists/ Optimised Files

Current State:
  Total files in optimised folder: 6179
  Active site/species combinations: 211

Non-Active Lists:
  2023 combinations: 41
  2024 combinations: 63
  Combined unique combinations: 72

Overlap Analysis:
  Non-active combinations with files: 70
  Total files affected: 1247
  Non-active combinations without files: 2

Projected State After Removal:
  Files remaining: 4932
  Active combinations remaining: 170
  Combinations removed: 41

Data Loss Percentage:
  Files: 20.18%
  Site/species combinations: 19.43%


#### out put of the analyse:
    Comparison Analysis Report Non-Active Lists/ Optimised Files

    
        Current State:
        Total files in optimised folder: 6179
        Active site/species combinations: 211

        Non-Active Lists:
        2023 combinations: 41
        2024 combinations: 63
        Combined unique combinations: 72

        Overlap Analysis:
        Non-active combinations with files: 70
        Total files affected: 1247
        Non-active combinations without files: 2

        Projected State After Removal:
        Files remaining: 4932
        Active combinations remaining: 170
        Combinations removed: 41

        Data Loss Percentage:
        Files: 20.18%
        Site/species combinations: 19.43%`

#### Below another function for calculating projected issue rate, if this removal happens.
    func called: calculate_projected_issue_rate.

In [5]:
def calculate_projected_issue_rate(comparison_results, optimised_root):
    """
    Calculates what the issue rate would be if non-active files were excluded from the calculation. Does not modify any files.
    
    Args:
        comparison_results: dict returned from compare_nonactive_with_files.
        optimised_root: Root directory containing optimised CSV files.
        
    Returns:
        dict with current and projected issue rates.
    """
    
    overlapping = comparison_results['overlapping_combinations']
    files_to_exclude = set()
    for files in overlapping.values():
        files_to_exclude.update(files)
    
    all_files = list(Path(optimised_root).rglob("*.csv"))
    
    # Calculate current issue rate ( will be %26 issue rate)
    current_issues = 0
    current_total = 0
    
    # Calculate projected issue rate excluding nonActive
    projected_issues = 0
    projected_total = 0
    
    for filepath in all_files:
        try:
            df = pd.read_csv(filepath)
            if '@Value' in df.columns and len(df) > 0:
                missing_count = df['@Value'].isna().sum()
                missing_percentage = (missing_count / len(df)) * 100
                
                current_total += 1
                if missing_percentage > 20:
                    current_issues += 1
                
                if filepath.name not in files_to_exclude:
                    projected_total += 1
                    if missing_percentage > 20:
                        projected_issues += 1
        except Exception as e:
            print(f"Error reading {filepath.name}: {e}")
    
    current_rate = (current_issues / current_total * 100) if current_total > 0 else 0.0
    projected_rate = (projected_issues / projected_total * 100) if projected_total > 0 else 0.0
    
    print("\nIssue Rate Comparison")
    print("=" * 70)
    print(f"\nCurrent Issue Rate (all files):")
    print(f"  Files checked: {current_total}")
    print(f"  Files with >20% missing: {current_issues}")
    print(f"  Issue rate: {current_rate:.2f}%")
    
    print(f"\nProjected Issue Rate (excluding non-active):")
    print(f"  Files checked: {projected_total}")
    print(f"  Files with >20% missing: {projected_issues}")
    print(f"  Projected issue rate: {projected_rate:.2f}%")
    
    print(f"\nImprovement:")
    improvement = current_rate - projected_rate
    print(f"  Issue rate reduction: {improvement:.2f} percentage points")
    
    return {
        'current_issue_rate': current_rate,
        'current_total': current_total,
        'current_issues': current_issues,
        'projected_issue_rate': projected_rate,
        'projected_total': projected_total,
        'projected_issues': projected_issues,
        'improvement': improvement
    }

In [6]:
# run projected issue rate calculation
projected_issue_stats = calculate_projected_issue_rate(result, optimased_root)


Issue Rate Comparison

Current Issue Rate (all files):
  Files checked: 6179
  Files with >20% missing: 1606
  Issue rate: 25.99%

Projected Issue Rate (excluding non-active):
  Files checked: 4932
  Files with >20% missing: 849
  Projected issue rate: 17.21%

Improvement:
  Issue rate reduction: 8.78 percentage points


#### If I optimase the removal according to 2023/2024 and 2025 my issue rate will be reduce %8.78 more. But at the same time number of the files will be also decrease.
    Issue Rate Comparison
    ======================================================================

    Current Issue Rate (all files):
    Files checked: 6179
    Files with >20% missing: 1606
    Issue rate: 25.99%

    Projected Issue Rate (excluding non-active):
    Files checked: 4932
    Files with >20% missing: 849
    Projected issue rate: 17.21%

    Improvement:
    Issue rate reduction: 8.78 percentage points.
 - At this stage, removal seems to be the better option for me because it will decrease the amount of unusable data. However, it will also cause me to lose some healthy files (those with 100% valid values in the value column). As a final step, I will create a function to compare the percentage of healthy versus unhealthy data, and then decide how to proceed.

## 5) Removal Quality Analysis

- Creating function called: analyse_removal_quality, 
    - What the function does? 
        - Evaluates whether removing non-active site/species files is justified by comparing the quality of files that would be removed. It categorises files into healthy (low missing values), unhealthy (high missing values), and perfect (no missing values).

    - Result: 
        - The analysis helps determine if the removal would sacrifice too many good quality files or if it effectively targets problematic data. A benefit ratio is calculated to show the trade-off between removing healthy versus unhealthy files.


In [7]:
def analyse_removal_quality(comparison_results, optimised_root, healthy_threshold=20):
    """
    Analyses the quality of files that would be removed by comparing healthy files vs unhealthy files. Helps determine if removal is justified or if too many good files would be lost.
    
    Args:
        comparison_results: dict from compare_nonactive_with_files
        optimised_root: root directory containing optimised CSV files
        healthy_threshold: percentage threshold for missing values (default 20%)
        
    Returns:
        dict containing removal quality metrics
    """
    
    overlapping = comparison_results['overlapping_combinations']
    files_to_remove = set()
    for files in overlapping.values():
        files_to_remove.update(files)
    
    all_files = list(Path(optimised_root).rglob("*.csv"))
    
    # categorise files to be removed
    healthy_files_removed = []
    unhealthy_files_removed = []
    perfect_files_removed = []  # 100% complete, 0% missing
    
    for filepath in all_files:
        if filepath.name in files_to_remove:
            try:
                df = pd.read_csv(filepath)
                if '@Value' in df.columns and len(df) > 0:
                    missing_count = df['@Value'].isna().sum()
                    missing_percentage = (missing_count / len(df)) * 100
                    
                    file_info = {
                        'filename': filepath.name,
                        'path': str(filepath),
                        'total_rows': len(df),
                        'missing_values': missing_count,
                        'missing_percentage': round(missing_percentage, 2)
                    }
                    
                    if missing_percentage == 0:
                        perfect_files_removed.append(file_info)
                        healthy_files_removed.append(file_info)
                    elif missing_percentage <= healthy_threshold:
                        healthy_files_removed.append(file_info)
                    else:
                        unhealthy_files_removed.append(file_info)
            except Exception as e:
                print(f"error reading {filepath.name}: {e}")
    
    total_files_to_remove = len(healthy_files_removed) + len(unhealthy_files_removed)
    healthy_count = len(healthy_files_removed)
    unhealthy_count = len(unhealthy_files_removed)
    perfect_count = len(perfect_files_removed)
    
    healthy_percentage = (healthy_count / total_files_to_remove * 100) if total_files_to_remove > 0 else 0
    unhealthy_percentage = (unhealthy_count / total_files_to_remove * 100) if total_files_to_remove > 0 else 0
    perfect_percentage = (perfect_count / total_files_to_remove * 100) if total_files_to_remove > 0 else 0
    
    # calculate worth assessment
    print("\nremoval quality analysis")
    print("=" * 70)
    
    print(f"\nfiles that would be removed:")
    print(f"  total files to remove: {total_files_to_remove}")
    print(f"  healthy files (<={healthy_threshold}% missing): {healthy_count} ({healthy_percentage:.2f}%)")
    print(f"  unhealthy files (>{healthy_threshold}% missing): {unhealthy_count} ({unhealthy_percentage:.2f}%)")
    print(f"  perfect files (0% missing): {perfect_count} ({perfect_percentage:.2f}%)")
    
    print(f"\nremoval worth assessment:")
    if healthy_percentage > unhealthy_percentage:
        print(f"  warning: more healthy files ({healthy_percentage:.2f}%) than unhealthy ({unhealthy_percentage:.2f}%) would be removed")
        recommendation = "not recommended"
    else:
        print(f"  removing more unhealthy files ({unhealthy_percentage:.2f}%) than healthy ({healthy_percentage:.2f}%)")
        recommendation = "recommended"
    
    print(f"\n  recommendation: {recommendation}")
    
    if perfect_count > 0:
        print(f"\n  note: {perfect_count} perfect files with complete data would be lost")
    
    # calculate benefit ratio
    benefit_ratio = unhealthy_count / healthy_count if healthy_count > 0 else float('inf')
    print(f"\n  benefit ratio (unhealthy/healthy): {benefit_ratio:.2f}")
    if benefit_ratio < 1:
        print(f"  removing 1 unhealthy file costs {1/benefit_ratio:.2f} healthy files")
    else:
        print(f"  removing 1 healthy file removes {benefit_ratio:.2f} unhealthy files")
    
    return {
        'total_files_to_remove': total_files_to_remove,
        'healthy_files_removed': healthy_files_removed,
        'unhealthy_files_removed': unhealthy_files_removed,
        'perfect_files_removed': perfect_files_removed,
        'healthy_count': healthy_count,
        'unhealthy_count': unhealthy_count,
        'perfect_count': perfect_count,
        'healthy_percentage': healthy_percentage,
        'unhealthy_percentage': unhealthy_percentage,
        'perfect_percentage': perfect_percentage,
        'benefit_ratio': benefit_ratio,
        'recommendation': recommendation
    }

In [8]:
# Run removal quality analysis
removal_quality_stats = analyse_removal_quality(result, optimased_root)


removal quality analysis

files that would be removed:
  total files to remove: 1247
  healthy files (<=20% missing): 490 (39.29%)
  unhealthy files (>20% missing): 757 (60.71%)
  perfect files (0% missing): 83 (6.66%)

removal worth assessment:
  removing more unhealthy files (60.71%) than healthy (39.29%)

  recommendation: recommended

  note: 83 perfect files with complete data would be lost

  benefit ratio (unhealthy/healthy): 1.54
  removing 1 healthy file removes 1.54 unhealthy files
